File: python/hot100/56.子集.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        ans = []
        current = []
        n = len(nums)
        def dfs(index) -> None:
            if index == n:
                ans.append(current.copy())
                return 
            dfs(index + 1)
            current.append(nums[index])
            dfs(index + 1)
            current.pop()
        dfs(0)
        return ans

File: cpp/hot100/56.子集.cpp
class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        int n = nums.size();
        vector<vector<int>> ans;
        vector<int> current;
        auto dfs = [&](this auto&& dfs, int index) -> void{
            if(index == n){
                ans.emplace_back(current);
                return ;
            }
            dfs(index + 1);
            current.emplace_back(nums[index]);
            dfs(index + 1);
            current.pop_back();
        };
        dfs(0);
        return ans;
    }
};